File: src/ct/game/input_events.c
#include <SDL.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <math.h>
#include "ct_stacktrace.h"
#include "input_events.h"
#include "engine.h"
#include "image.h"
#include "utils.h"
#include "image.h"
#include "ct_array.h"
#include "game.h"

void void_mfunc_InputEvents_update_InputEventsPtr_GamePtr_bool(CT_StackTrace* ct_stacktrace, int func_idx_of_call, int line_number_of_call, InputEvents* self, Game* game, bool _is_mouse_over) {
    ct_stacktrace_push(ct_stacktrace, func_idx_of_call, line_number_of_call);
    self->is_click = false;
    if (_is_mouse_over) {
        if (!(self->is_mouse_over)) {
            self->is_mouse_enter = true;
}
        self->is_mouse_over = true;
}
    else {
        if (self->is_mouse_over) {
            self->is_mouse_exit = false;
}
        self->is_mouse_over = false;
}
    if (game->engine.is_mouse_down) {
        if (self->is_mouse_over) {
            self->was_mouse_down_when_mouse_over = true;
            self->is_mouse_down = true;
}
        else {
            self->was_mouse_down_when_mouse_over = false;
    }
}
    if (game->engine.is_mouse_up) {
        if (self->is_mouse_over) {
            self->is_mouse_up = true;
            if (self->was_mouse_down_when_mouse_over) {
                self->is_click = true;
    }
}
        self->was_mouse_down_when_mouse_over = false;
}
    if (game->engine.is_mouse_held_down) {
        if (self->is_mouse_over) {
            self->is_mouse_held_down = true;
    }
}
    if (game->engine.is_right_mouse_down) {
        if (self->is_mouse_over) {
            self->was_right_mouse_down_when_mouse_over = true;
            self->is_right_mouse_down = true;
}
        else {
            self->was_right_mouse_down_when_mouse_over = false;
    }
}
    if (game->engine.is_right_mouse_up) {
        if (self->is_mouse_over) {
            self->is_right_mouse_up = true;
            if (self->was_right_mouse_down_when_mouse_over) {
                self->is_right_click = true;
    }
}
        self->was_right_mouse_down_when_mouse_over = false;
}
    if (game->engine.is_right_mouse_held_down) {
        if (self->is_mouse_over) {
            self->is_right_mouse_held_down = true;
        }
    }
    ct_stacktrace_pop(ct_stacktrace);
}

void void_mfunc_InputEvents_clear_InputEventsPtr(CT_StackTrace* ct_stacktrace, int func_idx_of_call, int line_number_of_call, InputEvents* self) {
    ct_stacktrace_push(ct_stacktrace, func_idx_of_call, line_number_of_call);
    self->is_mouse_enter = false;
    self->is_mouse_exit = false;
    self->is_mouse_over = false;
    self->is_mouse_down = false;
    self->is_mouse_held_down = false;
    self->is_mouse_up = false;
    self->is_right_mouse_down = false;
    self->is_right_mouse_held_down = false;
    self->is_right_mouse_up = false;
    self->is_click = false;
    self->is_right_click = false;
    self->was_mouse_down_when_mouse_over = false;
    self->was_right_mouse_down_when_mouse_over = false;
    ct_stacktrace_pop(ct_stacktrace);
}

File: src/ct/src/input_events.py
from ct.src.engine import SDL_RenderCopy
from ct.src.image import *
from ct.src.utils import SDL_Rect, RenderDst, set_scaled_rect, set_scaled_rect_with_camera
from ct.src.image import Image
from ct.src.ct_array import Array
from ct.src.base_types import addr, Uint32, mod
# weak imports
from ct.src.game import Game


class InputEvents:
    def __init__(self):
        self.is_mouse_enter: bool = None
        self.is_mouse_exit: bool = None
        self.is_mouse_over: bool = None
        self.is_mouse_down: bool = None
        self.is_mouse_held_down: bool = None
        self.is_mouse_up: bool = None
        self.is_right_mouse_down: bool = None
        self.is_right_mouse_held_down: bool = None
        self.is_right_mouse_up: bool = None
        self.is_click: bool = None
        self.is_right_click: bool = None
        self.was_mouse_down_when_mouse_over: bool = None
        self.was_right_mouse_down_when_mouse_over: bool = None

    def clear(self):
        self.is_mouse_enter = False
        self.is_mouse_exit = False
        self.is_mouse_over = False
        self.is_mouse_down = False
        self.is_mouse_held_down = False
        self.is_mouse_up = False
        self.is_right_mouse_down = False
        self.is_right_mouse_held_down = False
        self.is_right_mouse_up = False
        self.is_click = False
        self.is_right_click = False
        self.was_mouse_down_when_mouse_over = False
        self.was_right_mouse_down_when_mouse_over = False

    def update(self, game: Game, _is_mouse_over: bool):
        self.is_click = False
        if _is_mouse_over:
            # call onMouseEnter once by calling it before the isMouseOver flag is set
            if not(self.is_mouse_over):
                self.is_mouse_enter = True
            self.is_mouse_over = True
        else:
            # call only once before flag is set
            if self.is_mouse_over:
                self.is_mouse_exit = False
            self.is_mouse_over = False
        # events that require mouse enter status
        # mouse down
        if game.engine.is_mouse_down:
            if self.is_mouse_over:
                self.was_mouse_down_when_mouse_over = True
                self.is_mouse_down = True
            else:
                self.was_mouse_down_when_mouse_over = False
        # mouse up
        if game.engine.is_mouse_up:
            if self.is_mouse_over:
                self.is_mouse_up = True
                if self.was_mouse_down_when_mouse_over:
                    self.is_click = True
            # always set to False on mouse up
            self.was_mouse_down_when_mouse_over = False
        # mouse held down
        if game.engine.is_mouse_held_down:
            if self.is_mouse_over:
                self.is_mouse_held_down = True
        # right mouse down
        if game.engine.is_right_mouse_down:
            if self.is_mouse_over:
                self.was_right_mouse_down_when_mouse_over = True
                self.is_right_mouse_down = True
            else:
                self.was_right_mouse_down_when_mouse_over = False
        # right mouse up
        if game.engine.is_right_mouse_up:
            if self.is_mouse_over:
                self.is_right_mouse_up = True
                if self.was_right_mouse_down_when_mouse_over:
                    self.is_right_click = True
            # always set to False on mouse up
            self.was_right_mouse_down_when_mouse_over = False
        # right mouse held down
        if game.engine.is_right_mouse_held_down:
            if self.is_mouse_over:
                self.is_right_mouse_held_down = True
